### PR TITLE
breaking: return a promise.

### PR DESCRIPTION
**Pull request by jamestalmage**
Includes major API change discussed in #8.

### CODE REVIEW

**Comment by SBoudrias** [comment]
`setPromise` is not something we can do. Considering the way npm is working, there's no way we can't guarantee the type of promise we'll return as multiple users might each overwrite it.

I don't really want to add a factory to generate a new run-async function... Maybe we should just drop this one until Node >= 4.0 is the norm.

**Comment by jamestalmage** [comment]
@SBoudrias 
OK, check the latest implementation. It just detects if there is a global `Promise` variable.

I skip the last two tests for the Node `0.10` build (using feature detection). My only additional thought is that we should maybe throw a helpful error if they fail to provide `cb` on Node `0.10`.

I'd like to see us start returning a promise now rather than waiting, I want to start using `yield runAsync`!

**Comment by jamestalmage** [comment]
Second commit adds a helpful error if there is no global Promise implementation. Should help prevent confusion for code that works in 0.12+, but not in 0.10

File: index.js [hunk 1]
```diff
@@ -12,25 +13,37 @@ var isPromise = require('is-promise');
...
+  return function () {
+    var async = false;
+    var promise = null;
+    cb = cb && once(cb);
+    if (typeof Promise !== 'undefined') {
+      promise = new Promise(function (resolve, reject) {
+        cb = promiseResolver(resolve, reject, cb);
+      });
+    } else if (!cb) {
+      throw new Error('no callback or global Promise implementation');
```
Comment (SBoudrias) [inline index.js]: Let's make this cleaner: "You need to provide a callback unless you're running on Node >= 4.0"

Comment (jamestalmage) [inline index.js]: 12.7 also works (not sure on 12.0-12.6).

You could also make it work on 0.10 by setting global.Promise to bluebird, etc.

How about:
"You need to provide a callback or upgrade to Node 0.12.?? if you intend to use Promises" 

We could discuss hacking 0.10 to work in the documentation

Comment (SBoudrias) [inline index.js]: Well, the main point for me is to explain what to do in the error message rather than what went wrong.

Comment (jamestalmage) [inline index.js]: Yep, makes sense. I can update shortly.

Do you think adding directions about setting global.promise seems overly verbose?

Comment (SBoudrias) [inline index.js]: Oh yes, let's keep it simple. I just want to change the focus of the current error message.

**Comment by jamestalmage** [comment]
`No Native Promise Implementation: You must upgrade to Node >= 0.11.13, or use callbacks`

I think that communicates all the possible solutions, while staying as short as possible.

**Comment by SBoudrias** [comment]
Perfect!!

**Pull request merged**